includes/class-wp-document-revisions-admin.php

```diff
@@ -1462,45 +1651,42 @@ public function delete_attachments_with_document( $post_id ) {
 			return;
 		}
 
-		if ( is_numeric( $document->post_content ) ) {
-			// check that content points to an attachment post.
-			$attach = get_post( $document->post_content );
-			if ( isset( $attach->ID ) && 'attachment' === $attach->post_type ) {
-
-				// make sure that the attachment is not refered to by another post (ignore autosave).
-				global $wpdb;
-				// phpcs:disable WordPress.DB.DirectDatabaseQuery
-				$doc_link = $wpdb->get_var(
-					$wpdb->prepare(
-						"SELECT COUNT(1) FROM $wpdb->posts WHERE %d IN (post_parent, id) AND post_content = %d AND post_name != %s ",
-						$document->post_parent,
-						$document->post_content,
-						strval( $document->post_parent ) . '-autosave-v1'
-					)
-				);
-				// phpcs:enable WordPress.DB.DirectDatabaseQuery
-
-				if ( '1' === $doc_link ) {
-					// look for attachment meta (before deleting attachment).
-					$meta = get_post_meta( $attach->ID, '_wp_attachment_metadata', true );
-
-					$file = get_attached_file( $attach->ID );
-					wp_delete_attachment( $attach->ID, false );
+		$attach = $this->get_document( $post_id );
+		if ( $attach ) {
+			// make sure that the attachment is not refered to by another post (ignore autosave).
+			global $wpdb;
+			// phpcs:disable WordPress.DB.DirectDatabaseQuery
+			$doc_link = $wpdb->get_var(
```

Comment:
Curious, why does this have to be done via a direct DB query?

Reply:
There are 3 parts to the where condition:
a)  %d IN (post_parent, id)
You can ask for those with post_parent = X or id = X - but not in one query
b) (post_content = %d OR post_content LIKE %s )
Can't put post_content as a WP_Query parameter 
c) post_name != %s
No negative test on name field. In fact, this is actually a status selection.
It is possible that a WP_Query would not retrieve auto-save status posts, but needs to be tested.

For me, to replace this by WP_Query would be:
a) get_post(document)
if condition met, then a match. (Can bypass b))
b) WP_Query on revisions (post_parent)
loop though each testing for condition.

So it is possible. But it is a load of code.